### ISSUE TITLE
replacing modified terminals

### ISSUE THREAD

**Issue by colinjcotter** [firedrakeproject/ufl#9]
In the Gusto equations rewrite project we are bumping into an issue with replacing components of mixed coefficients/test functions.

The setup is that we want to write down a UFL form describing a mixed problem, and extract the subproblem for one of the components. The code that describes this is:

from firedrake import *

mesh = UnitSquareMesh(10,10)
V = FunctionSpace(mesh,"CG",1)
W = V * V

w1, w2 = TestFunctions(W)

u = Function(W)

u1, u2 = split(u)

F = (u1*w1 + u2*w2)*dx

w = TestFunction(V)

u0 = Function(V)

F1 = replace(F, {w1:0})
F2 = replace(F1, {w2:w})
F3 = replace(F2, {u2,u0})

This does not work because w1 is not a terminal, it is an indexed mixed test function. David
says that w1 is a modified terminal.

What is the solution to this?

**Comment by dham** [comment]
I think that the extended replace on the `extended_replace` UFL branch fixes this. I prefer that to rolling our own second replacer class.

**Comment by wence-** [comment]
Your solution doesn't work for the case:
```python
mesh = UnitSquareMesh(1, 1)
V = FunctionSpace(mesh, "RT", 1)
Q = FunctionSpace(mesh, "P", 1)

W = V*Q
v, q = TestFunctions(W)

v2 = TestFunction(V)

L = inner(as_vector([1, 1]), v)*dx + q*dx

L2 = replace(L, {v: v2})
```

Because `v` is not an `Indexed`, but rather a `ListTensor`. I would advocate that we instead try and see if anything breaks anywhere if we replace the UFL implementation with my proposal.

**Comment by dham** [comment]
I'm nervous about putting a blanket replacer in. If that works, why did Martin not do that to start with? I'm concerned that there will be other cases that don't work that we're not thinking about.

**Comment by wence-** [comment]
I agree with that nervousness. How many different cases do we want to support? Is it just `Terminal`, `Indexed`, and `ListTensor`?

**Comment by dham** [comment]
Those are the ones I can think of

**Comment by thomasgibson** [comment]
How about going the other way: `V*Q` to say `V*Q*W`? Does your extended replacer handle the case where you expand the size of the mixed space?

**Comment by colinjcotter** [comment]
It looks like the modified terminal can be replaced by anything, including indexed test functions and coefficients that are from a different mixed FS.


all the best

--cjc

________________________________

**Comment by dham** [comment]
Yes, what you substitute in can be anything. It's what you substitute for which is hairy

**Comment by dham** [comment]
Yesish. Basically yes, but think carefully about the checks

Sent from my iPhone

On 27 Nov 2018, at 11:23, Colin J Cotter <notifications@github.com<mailto:notifications@github.com>> wrote:


So what's the plan with this? Do we extend extended_replace further to include all three of those types?

**Comment by jshipton** [comment]
I can't seem to make an LVP on a mixed function space using this branch of UFL.

```
from firedrake import *
import ufl

# set up mesh and function spaces
mesh = UnitSquareMesh(2, 2)
Vcg = FunctionSpace(mesh, "CG", 1)
Vdg = FunctionSpace(mesh, "DG", 1)
W = MixedFunctionSpace((Vcg, Vdg))

test1, test2 = TestFunctions(W)
tri1, tri2 = TrialFunctions(W)
f1, f2 = Function(W).split()

a = test1*tri1*dx + test2*tri2*dx
L = test1*f1*dx + test2*f2*dx
f = Function(W)
prob = LinearVariationalProblem(a, L, f)
```

Gives:

Traceback (most recent call last):
  File "test_replace2.py", line 17, in <module>
    prob = LinearVariationalProblem(a, L, f)
  File "/Users/jshipton/vanilla_firedrake/src/firedrake/firedrake/variational_solver.py", line 280, in __init__
    form_compiler_parameters=form_compiler_parameters)
  File "/Users/jshipton/vanilla_firedrake/src/firedrake/firedrake/variational_solver.py", line 48, in __init__
    raise ValueError("Provided residual is not a linear form")
ValueError: Provided residual is not a linear form

**Comment by dorugeber** [comment]
`f1, f2 = split(Function(W))`?

**Comment by dorugeber** [comment]
Unless you're also in complex, in which case you need `inner(...)` rather than `*` to handle conjugation etc.

**Comment by jshipton** [comment]
Sorry, I should've mentioned that exactly this code works with the master branch of UFL.

`split` instead of `.split()` makes no difference and not using complex.

**Comment by jshipton** [comment]
[Although possibly I should be using `split` anyway... thanks @dorugeber]

**Comment by dham** [comment]
OK, so what’s going wrong in the replace?

**Comment by jshipton** [comment]
I wish I knew! I can tell you that if I comment out the line `indexed = terminal` then the code works (revealing my sophisticated debugging strategies now!). I can't track down where `indexed` is used... I'm flailing around in the depths of UFL now with not much clue about what is going on!

**Comment by colinjcotter** [comment]
You could have a look at what is produced comparing with and without that line.  If you print the UFL form then it is a mess but you might spot the difference!
________________________________

**Comment by jshipton** [comment]
The difference is that the replace hasn't done anything - the trial functions have not been replaced with `u`. That's why `F` is not a linear form - it still has 2 arguments. But I don't understand why - I don't know what that line `indexed = terminal` is for.

**Comment by wence-** [comment]
So `action(a, u)` does something like:

```python
test, trial = a.arguments()

au = replace(a, {test: u})
```

But in this mixed form, the trial function is `ListTensor([Indexed(trial, 0), Indexed(trial, 1)])` or similar.

By treating `indexed` like `terminal`, we try and match `Indexed(trial, 0)` etc... against our replacement dict but it doesn't, and so we just return the node untouched.

I think that for non-terminals that we want to replace, if we don't match, we have to traverse into the children and reconstruct.

**Comment by wence-** [comment]
Doing this in the implementation is like what I posted above, but instead of providing a handler that does this for all nodes (via `expr`) we specifically enumerate the nodes we want to treat (which feels fragile).

**Comment by dham** [comment]
Hmmm. Maybe we should put in a general replacer and be damned. The fact that arguments in mixed spaces are not actually terminals is going to make any attempt to constrain what we replace pretty dubious.

**Comment by wence-** [comment]
I'll propose it as a PR and test to see if we break anything.

**Comment by colinjcotter** [comment]
OK thanks Lawrence! We'll be ready with some tests.

________________________________

**Comment by jshipton** [comment]
I've made a branch `generic_replace` with Lawrence's suggested code and I'm happily using it in the `programmable_eqn_transformations` branch of Gusto. Do you have any suggestions for tests or sanity checks?

**Comment by wence-** [comment]
This was fixed in FEniCS/ufl#15.
